### PR TITLE
Halo2 aggregation

### PR DESCRIPTION
**Pull request by leonardoalt**
This is based on the examples in https://github.com/privacy-scaling-explorations/snark-verifier/

### CODE REVIEW

File: halo2/src/aggregation.rs [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+    },
+    util::arithmetic::{fe_to_limbs, PrimeField},
+    verifier::{self, plonk::PlonkProtocol, SnarkVerifier},
+};
+
+use itertools::Itertools;
+use rand::rngs::OsRng;
+use std::rc::Rc;
+
+const LIMBS: usize = 4;
```
Comment (Schaeff) [inline halo2/src/aggregation.rs:44]: Would be nice to have a short comment explaining what these are

**Review by Schaeff — COMMENTED** [review]
Really nice, any chance we can have some tests though?

File: halo2/src/prover.rs [hunk 2]
```diff
@@ -28,10 +38,6 @@ pub fn prove_ast_read_params<T: FieldElement, R: io::Read>(
     witness: Vec<(&str, Vec<T>)>,
     mut params: R,
 ) -> Vec<u8> {
-    if polyexen::expr::get_field_p::<Fr>() != T::modulus().to_arbitrary_integer() {
```
[the remark this answers is not in the document]
Comment (leonardoalt) [reply halo2/src/prover.rs:31]: yea in the next function that's called in here

File: halo2/src/prover.rs [hunk 3]
```diff
@@ -50,42 +56,101 @@ pub fn prove_ast<T: FieldElement>(
...
+
+    let circuit_app = analyzed_to_circuit(pil, fixed, witness);
+
+    log::debug!("{}", PlafDisplayBaseTOML(&circuit_app.plaf));
+
+    let vk_app = keygen_vk(&params_app, &circuit_app).unwrap();
+    let protocol_app = compile(
+        &params_app,
+        &vk_app,
+        Config::kzg().with_num_instance(vec![]),
```
[the remark this answers is not in the document]
Comment (leonardoalt) [reply halo2/src/prover.rs:126]: yea

**Comment by leonardoalt** [comment]
> Really nice, any chance we can have some tests though?

Tests are not be possible in the current CI setup because the aggr circuit requires 64gb ram in the prover. We should anyway soon get a self-hosted runner so we can do this kind of stuff.

File: halo2/src/aggregation.rs (hunk 1 above)
Comment (leonardoalt) [reply halo2/src/aggregation.rs:51]: added

**Comment by leonardoalt** [comment]
Fixed review comments and rebased.

**Review by Schaeff — APPROVED** [review]

**Pull request merged**